### PR TITLE
change tiles.py to avoid rendering empty tiles.

### PR DESCRIPTION
**Pull request by abayuen**
All tiles within the full_data_extent were rendered. However, not all tiles contained data and many of them can be empty.
In this PR, we added check to avoid rendering empty tiles.

### CODE REVIEW

**Comment by jbednar** [comment]
Do all tile-server clients reliably handle missing tiles, e.g. by scaling the tile from the previous zoom level?  If not it doesn't seem safe to skip the empty tiles in general, though it would make sense to add an option for it.

**Comment by abayuen** [comment]
Hello @jbednar,
Thanks for your contribution to datashader!

All time-server client should reliable handle 'missing' tiles. Datashader is already introducing 'missing tiles' by passing full_extent as argument in the render_tile function, in a sense that the tiles outside of full_extend are not generated.

**Comment by abayuen** [comment]
I am not sure what the rules are for merging to master.. so I will wait for your reply

**Comment by jbednar** [comment]
> Datashader is already introducing 'missing tiles' by passing full_extent as argument in the render_tile function, in a sense that the tiles outside of full_extend are not generated.

I think that's a different issue.  Sure, all sets of tiles are finite, and so unless they cover the world, it's possible to come to the end of them.  But that's different than saying that at some zoom level, a particular tile in the middle that has no data should not be rendered at all; in some cases it should, and others not.  So I think there are valid reasons for wanting a tile to be generated even when there is no data, so I'm happy to merge this once there is an argument to turn this behavior on and off.

**Comment by brendancol** [comment]
@abayuen I'm not sure if the `sum == 0` is a good-enough heuristic for empty tiles...but maybe it would...I'm looking at this soon and will ping back

**Comment by jbednar** [comment]
@brendancol, did you ever have a check about whether these changes are ok?  They seem fine to me if they are controlled by an option, but I'm less sure they should be the default.

File: datashader/tiles.py [hunk 1]
```diff
@@ -92,7 +94,11 @@ def render_super_tile(tile_info, output_path, load_data_func,
     agg = rasterize_func(df, x_range=tile_info['x_range'],
                          y_range=tile_info['y_range'], height=tile_size,
                          width=tile_size)
-    ds_img = shader_func(agg, span=tile_info['span'])
+    # if the rasterized data is empty, do not create subtiles.
+    if np.sum(agg) == 0:
+        pass
```
Comment (philippjfr) [inline datashader/tiles.py]: Won't this just error because `ds_img` won't be defined below?

Comment (jbednar) [reply datashader/tiles.py]:
```suggestion
        return
```
?

Comment (brendancol) [reply datashader/tiles.py]: @jbednar I would vote for closing this PR and not using the `np.sum(agg) == 0` heuristic because it feels brittle to me.

Comment (jbednar) [reply datashader/tiles.py]: Ok, will do.  Anyone who feels otherwise can reopen while addressing the above issue(s).

**Pull request closed without merging**